packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart

```diff
@@ -0,0 +1,411 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+
+import '../../globals.dart';
+import '../../service_extensions.dart';
+/* Dead imports to uncomment when the rest of the file is ported.
+import '../../service_manager.dart' show ServiceExtensionState;
+import '../../service_registrations.dart';
+import '../../utils.dart';
+*/
+
+// TODO(jacobr): add support for analytics.
+//import '../analytics.dart' as ga;
+import '../../service_registrations.dart';
+import '../../utils.dart';
+import 'flutter_icon_renderer.dart';
+
+class ServiceExtensionButtonGroup extends StatefulWidget {
+  const ServiceExtensionButtonGroup({
+    this.minIncludeLabelWidth,
+    @required this.extensions,
+  });
+
+  final double minIncludeLabelWidth;
+  final List<ToggleableServiceExtensionDescription> extensions;
+
+  @override
+  _ServiceExtensionButtonGroupState createState() =>
+      _ServiceExtensionButtonGroupState();
+}
+
+/// Data class tracking the state of a single service extension.
+class ExtensionState {
+  ExtensionState(this.description);
+
+  final ToggleableServiceExtensionDescription description;
+  bool isSelected = false;
+  bool isAvailable = false;
+  StreamSubscription subscription;
+
+  void dispose() {
+    subscription?.cancel();
+    subscription = null;
+  }
+}
+
+class _ServiceExtensionButtonGroupState
+    extends State<ServiceExtensionButtonGroup> {
+  List<ExtensionState> _extensionStates;
+
+  @override
+  void initState() {
+    super.initState();
+    _extensionStates = widget.extensions
+        .map((extension) => ExtensionState(extension))
+        .toList();
+    for (var extension in _extensionStates) {
+      final extensionName = extension.description.extension;
+      serviceManager.serviceExtensionManager
+          .getServiceExtensionState(extensionName, (state) {
+        setState(() {
+          extension.isSelected =
+              state.value == extension.description.enabledValue;
+        });
+      });
+      extension.isAvailable = serviceManager.serviceExtensionManager
+          .isServiceExtensionAvailable(extensionName);
+
+      extension.subscription =
+          serviceManager.serviceExtensionManager.hasServiceExtension(
+        extensionName,
+        (available) {
+          setState(() {
+            extension.isAvailable = available;
+          });
+        },
+      );
```

Comment:
do we need both of these? I would think that we wouldn't need to check `isServiceExtensionAvailable` if we are subscribing to `hasServiceExtension`

Reply:
I think that was a bug in service_extension_elements.dart
Verified and fixed.